File: Fondamenti di Informatica 2/Esercitazioni con soluzioni/8.2_DFS Diretta/Soluzione/c-source/graph_services.c
#include <stdio.h>

#include "graph.h"
#include "linked_list.h"
#include "graph_services.h"

static int sweep_aux(graph_node * source, int time, char * format_string) {
	if (source->status != UNEXPLORED)
		return 0;
	int loctime = 1;
	source->status = EXPLORING;
	source->timestamp = time;

	linked_list_iterator * lli = linked_list_iterator_new(source->out_edges);
	graph_node * u = (graph_node *)linked_list_iterator_getvalue(lli);
	while (linked_list_iterator_hasnext(lli)) {
		printf("\t");
		printf(format_string, source->value);
		printf("(%d)->", source->timestamp);
		printf(format_string, u->value);
		printf("(%d) ", u->timestamp);

		if (u->status == EXPLORED) {
			if (source->timestamp < u->timestamp)
				printf("FORWARD\n");
			else
				printf("CROSS\n");
		}
		else if (u->status == EXPLORING) {
			printf("BACK\n");
		}
		else {
			printf("TREE\n");
			loctime += sweep_aux(u, time + 1, format_string);
		}
		u = (graph_node*)linked_list_iterator_next(lli);
	}
	linked_list_iterator_delete(lli);
	source->status = EXPLORED;
	return loctime;
}

void sweep(graph * g, char * format_string) {
	linked_list_iterator * lli = linked_list_iterator_new(g->nodes);
	graph_node * n = (graph_node *)linked_list_iterator_getvalue(lli);
	while (linked_list_iterator_hasnext(lli)) {
		n->status = UNEXPLORED;
		n->timestamp = 0;
		n = (graph_node*)linked_list_iterator_next(lli);
	}
	linked_list_iterator_delete(lli);
	int loctime = 0;
	lli = linked_list_iterator_new(g->nodes);
	graph_node * node = (graph_node *)linked_list_iterator_getvalue(lli);
	while (linked_list_iterator_hasnext(lli)) {
		printf("Root ");
		printf(format_string, node->value);
		printf("\n");

		loctime += sweep_aux(node, loctime, format_string);
		node = (graph_node*)linked_list_iterator_next(lli);
	}
	linked_list_iterator_delete(lli);
}

static int DDFS(graph_node* node, linked_list* list) {
	if (node->status == EXPLORING)
		return 1;
	if (node->status == EXPLORED) {
		return 0;
	}
	node->status = EXPLORING;
	int ret = 0;
	linked_list_iterator* lli = linked_list_iterator_new(node->out_edges);
	graph_node* n = (graph_node*)linked_list_iterator_getvalue(lli);
	for (; linked_list_iterator_hasnext(lli); n = (graph_node*)linked_list_iterator_next(lli)) {
		ret += DDFS(n, list);
	}
	linked_list_iterator_delete(lli);
	node->status = EXPLORED;
	linked_list_add(list, 0 ,node);
	return ret;
}

void topological_sort(graph * g){
	linked_list_iterator * lli = linked_list_iterator_new(g->nodes);
	graph_node * n = (graph_node *)linked_list_iterator_getvalue(lli);
	while (linked_list_iterator_hasnext(lli)) {
		n->status = UNEXPLORED;
		n->timestamp = 0;
		n = (graph_node*)linked_list_iterator_next(lli);
	}
	linked_list_iterator_delete(lli);
	linked_list* list = linked_list_new();
	lli = linked_list_iterator_new(g->nodes);
	n = (graph_node*)linked_list_iterator_getvalue(lli);
	for (; linked_list_iterator_hasnext(lli); n = (graph_node*)linked_list_iterator_next(lli)) {
		if (n->status == UNEXPLORED) {
			if (DDFS(n, list)) {
				printf("Impossibile ottenere ordine topologico, il grafo non e' un DAG\n");
				linked_list_delete(list);
				linked_list_iterator_delete(lli);
				return;
			}
		}
	}
	linked_list_iterator_delete(lli);
	lli = linked_list_iterator_new(list);
	n = (graph_node*)linked_list_iterator_getvalue(lli);
	for (; linked_list_iterator_hasnext(lli); n = (graph_node*)linked_list_iterator_next(lli)) {
		printf("%s ", (char*)(n->value));
	}
	printf("\n");
	linked_list_delete(list);
	linked_list_iterator_delete(lli);
}

/* SOLUTION WITH ONLY OUT EDGES
static void transposed_DDFS(graph* g, graph_node * node, linked_list* list) {
	if (node->status != UNEXPLORED)
		return;
	node->status = EXPLORING;
	linked_list_iterator* it = linked_list_iterator_new(g->nodes);
	graph_node* to = (graph_node*)linked_list_iterator_getvalue(it);
	for (; linked_list_iterator_hasnext(it); to = linked_list_iterator_next(it)) {
		linked_list_iterator* it2 = linked_list_iterator_new(to->out_edges);
		graph_node* frm = (graph_node*)linked_list_iterator_getvalue(it2);
		for (; linked_list_iterator_hasnext(it2); frm = linked_list_iterator_next(it2)) {
			if (frm == node)
				transposed_DDFS(g, to, list);
		}
		linked_list_iterator_delete(it2);
	}
	linked_list_iterator_delete(it);
	linked_list_enqueue(list, node);
	node->status = EXPLORED;
}
*/

static void transposed_DDFS(graph_node* node, linked_list* list) {
	if (node->status != UNEXPLORED)
		return;
	node->status = EXPLORING;
	linked_list_iterator* lli = linked_list_iterator_new(node->in_edges);
	graph_node* n = (graph_node*)linked_list_iterator_getvalue(lli);
	for (; linked_list_iterator_hasnext(lli); n = (graph_node*)linked_list_iterator_next(lli)) {
		transposed_DDFS(n, list);
	}
	linked_list_iterator_delete(lli);
	node->status = EXPLORED;
	linked_list_enqueue(list, node);
}

void strong_connected_components(graph  *g) {
	// Reset status of nodes
	linked_list_iterator * lli = linked_list_iterator_new(g->nodes);
	graph_node * n = (graph_node *)linked_list_iterator_getvalue(lli);
	while (linked_list_iterator_hasnext(lli)) {
		n->status = UNEXPLORED;
		n->timestamp = 0;
		n = (graph_node*)linked_list_iterator_next(lli);
	}
	// Create the stack, first DFS
	linked_list* stack = linked_list_new();
	linked_list_iterator_delete(lli);
	lli = linked_list_iterator_new(g->nodes);
	n = (graph_node*)linked_list_iterator_getvalue(lli);
	for (; linked_list_iterator_hasnext(lli); n = (graph_node*)linked_list_iterator_next(lli)) {
		if (n->status == UNEXPLORED) {
			DDFS(n, stack);
		}
	}
	linked_list_iterator_delete(lli);
	// Reset status of nodes
	lli = linked_list_iterator_new(g->nodes);
	n = (graph_node *)linked_list_iterator_getvalue(lli);
	while (linked_list_iterator_hasnext(lli)) {
		n->status = UNEXPLORED;
		n->timestamp = 0;
		n = (graph_node*)linked_list_iterator_next(lli);
	}
	linked_list_iterator_delete(lli);
	// Second DFS on the transposed graph
	lli = linked_list_iterator_new(stack);
	n = (graph_node*)linked_list_iterator_getvalue(lli);
	for (; linked_list_iterator_hasnext(lli); n = (graph_node*)linked_list_iterator_next(lli)) {
		if (n->status == UNEXPLORED) {
			linked_list* ret = linked_list_new();
			transposed_DDFS(n, ret);
			linked_list_iterator *it2 = linked_list_iterator_new(ret);
			graph_node* cur = (graph_node*)linked_list_iterator_getvalue(it2);
			printf("Strong connected component:\n");
			for (; linked_list_iterator_hasnext(it2); cur = (graph_node*)linked_list_iterator_next(it2)) {
				printf("%s ", (char*)(cur->value));
			}
			printf("\n");
			linked_list_iterator_delete(it2);
			linked_list_delete(ret);
		}
	}
	linked_list_iterator_delete(lli);
	linked_list_delete(stack);
}

File: Fondamenti di Informatica 2/Esercitazioni con soluzioni/8.2_DFS Diretta/Soluzione/java-source/GraphServices.java
import java.util.LinkedList;
import java.util.List;

public class GraphServices<V> {

    public static <V> void sweep(Graph<V> g) {
        // Reset node's status
        g.resetStatus();
        
       
    }

    private static <V> void sweep_aux(Graph g, Graph.Node<V> node) {
        node.state = Graph.Node.Status.EXPLORING;
    }

    private static<V> int DDFS(Graph.Node<V> nd, LinkedList<Graph.Node<V>> ts) {
        if(nd.state == Graph.Node.Status.EXPLORING)
            return 1;
        
        if(nd.state == Graph.Node.Status.EXPLORED)
            return 0;
        
        nd.state = Graph.Node.Status.EXPLORING;
        int ret = 0;
        
        for(Graph.Node<V> cur : nd.outEdges)
            ret += DDFS(cur, ts);
        
        nd.state = Graph.Node.Status.EXPLORED;
        ts.addFirst(nd);
        return ret;
    }

    private static<V> void transposedDFS(Graph.Node<V> nd, LinkedList<Graph.Node<V>> ret) {
        if(nd.state == Graph.Node.Status.EXPLORING)
            return;
        
        if(nd.state == Graph.Node.Status.EXPLORED)
            return;
        
        nd.state = Graph.Node.Status.EXPLORING;
        for(Graph.Node<V> cur : nd.inEdges)
            transposedDFS(cur, ret);
        
        ret.addLast(nd);
        nd.state = Graph.Node.Status.EXPLORED;
    }

    /* SOLUTION ONLY WITH OUT EDGES
    private static<V> void transposedDFS(Graph<V> g, Graph.Node<V> n, LinkedList<Graph.Node<V>> ret) {
        if(n.state != Graph.Node.Status.UNEXPLORED)
            return;
        
        n.state = Graph.Node.Status.EXPLORING;
        for(Graph.Node<V> to : g.getNodes()) {
            for(Graph.Node<V> frm : to.outEdges) {
                if(frm == n)
                    transposedDFS(g, to, ret);
            }
        }
        
        ret.addLast(n);
        n.state = Graph.Node.Status.EXPLORED;
    }
    */

    // This is Kosaraju's algorithm
    public static <V> void strongConnectedComponents(Graph<V> g) {
        // Reset node's status
        g.resetStatus();

        // First DFS
        LinkedList<Graph.Node<V>> stack = new LinkedList<>();
        for(Graph.Node<V> n : g.getNodes()) {
            if(n.state == Graph.Node.Status.UNEXPLORED)
                DDFS(n, stack);
        }

        // Reset node's status
        g.resetStatus();

        // Second DFS on the transposed graph
        for(Graph.Node<V> n : stack) {
            if(n.state == Graph.Node.Status.UNEXPLORED) {
                LinkedList<Graph.Node<V>> ret = new LinkedList<>();
                transposedDFS(n, ret);
                
                System.out.println("Strong connected component:");
                for(Graph.Node<V> cur : ret)
                    System.out.print(cur.value + " ");
                System.out.println("");
            }
        }
    }

    public static <V> void topologicalSort(Graph<V> g) {
        // Reset node's status
        g.resetStatus();

        LinkedList<Graph.Node<V>> ts = new LinkedList<Graph.Node<V>>();
        
        for(Graph.Node<V> nd : g.getNodes()) {
            if(nd.state == Graph.Node.Status.UNEXPLORED) {
                if(DDFS(nd, ts) > 0) {
                    System.out.println("Impossibile ottenere ordine topologico, il grafo non è un DAG");
                    return;
                }
            }
        }

        for(Graph.Node<V> nd : ts)
            System.out.print(nd.value + " ");
        System.out.println("");
    }
}
